### PR TITLE
Add support for TCP tunnels

### PR DESCRIPTION
**Pull request by farmio**
## Description

Add support for TCP tunnels
Some refactoring in the `io` module

Fixes # (issue)

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Checklist:

- [ ] I have performed a self-review of my own code
- [ ] The documentation has been adjusted accordingly
- [ ] The changes generate no new warnings
- [ ] Tests have been added that prove the fix is effective or that the feature works
- [ ] The changes are documented in the changelog
- [ ] The Homeassistant plugin has been adjusted in case of new config options

### CODE REVIEW

File: xknx/io/tunnel.py [hunk 1]
```diff
@@ -380,10 +341,191 @@ def _disconnect_request_received(
...
+        auto_reconnect_wait: int = 3,
+    ):
+        """Initialize Tunnel class."""
+        self.gateway_ip = gateway_ip
+        self.gateway_port = gateway_port
+
+        super().__init__(
+            xknx=xknx,
+            telegram_received_callback=telegram_received_callback,
+            auto_reconnect=auto_reconnect,
```
Comment (marvin-w) [inline xknx/io/tunnel.py:370]: Did you test the auto reconnect stuff?

File: xknx/io/connection.py [hunk 2]
```diff
@@ -10,9 +10,10 @@
 class ConnectionType(Enum):
     """Enum class for different types of KNX/IP Connections."""
 
-    AUTOMATIC = 0
-    TUNNELING = 1
-    ROUTING = 2
+    AUTOMATIC = auto()
+    ROUTING = auto()
+    TUNNELING = auto()
+    TUNNELING_TCP = auto()
```
[the remark this answers is not in the document]
Comment (farmio) [reply xknx/io/connection.py:16]: Yes, in HA we just use a string and match against that here https://github.com/home-assistant/core/blob/849abaca8bbb61947bd7fa0f4923db8c466406fe/homeassistant/components/knx/__init__.py#L391

File: xknx/io/tunnel.py (hunk 1 above)
Comment (farmio) [reply xknx/io/tunnel.py:370]: no there are still no tests for this

File: xknx/io/interface.py [hunk 3]
```diff
@@ -5,6 +5,8 @@
 * It starts and stops a udp client
 * It packs Telegrams into KNX Frames and passes them to a udp client
 """
+from __future__ import annotations
```
[the remark this answers is not in the document]
Comment (farmio) [reply xknx/io/interface.py:8]: we don't, its a leftover

**Pull request merged**